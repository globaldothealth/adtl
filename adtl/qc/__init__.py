"""
Quality Control module for ADTL
"""
import math
import json
import functools
import copy
from pathlib import Path
from typing import List, TypedDict, Optional, Callable, Tuple, Dict, Any

import pandas as pd
import numpy as np
import fastjsonschema


class Rule(TypedDict):
    module: str
    name: str
    description: str
    long_description: Optional[str]
    pattern: str


class WorkUnit(TypedDict):
    rule: Rule
    file: str
    dataset: str


class Dataset(TypedDict):
    folder: str
    files: List[str]


class WorkUnitResult(TypedDict):
    rule: str
    dataset: str
    file: str
    rows_success: int
    rows_fail: int
    rows: int
    ratio_success: float
    rows_fail_idx: List[int]
    success: bool
    mostly: float
    reason: str
    fail_data: pd.DataFrame


def rules_for(pattern: str, *rules):
    for r in rules:
        r.pattern = pattern


def rule(columns: List[str], mostly: float = 0, set_missing_columns: bool = True):
    """Decorator that indicates a QC rule

    Args:
        columns: List of required columns in the dataframe. If columns are not
        present, they are assigned null.

        mostly: Indicates the check in the rule should be considered a success
        if the ratio of successful rows is greater than this value. If not
        specified all rows have to succeed for the rule to be a considered a
        success.

        set_missing_columns: On by default, this setting assigns NA/null to
        required columns that are missing in the data
    """

    def decorator_rule(func):
        @functools.wraps(func)
        def wrapper(df, **kwargs):
            if set_missing_columns:
                for c in set(columns) - set(df.columns):
                    df[c] = None
            series = func(df, **kwargs)
            assert len(series) == len(
                df
            ), "Returned series must have same cardinality as source dataframe"
            rows_fail_idx = [i for i, val in enumerate(series) if val is False]
            if isinstance(series, (pd.Series, np.ndarray)):
                rows_success: int = series.sum()
                rows_fail = len(series) - rows_success
                ratio_success = rows_success / len(series)
                return dict(
                    rows_success=int(rows_success),
                    rows_fail=int(rows_fail),
                    rows=int(rows_success) + int(rows_fail),
                    ratio_success=ratio_success,
                    success=bool(ratio_success >= mostly),
                    mostly=mostly,
                    rows_fail_idx=",".join(map(str, rows_fail_idx)),
                    fail_data=json.dumps(
                        df.loc[rows_fail_idx][columns].to_dict(),
                        sort_keys=True,
                    ),
                )
            elif isinstance(series, bool):
                return dict(
                    rows_success=None,
                    rows_fail=None,
                    rows=None,
                    ratio_success=None,
                    success=series,
                    mostly=mostly,
                    rows_fail_idx=None,
                    fail_data=None,
                )

        return wrapper

    return decorator_rule


def _get_columns_from_valid_data(df: pd.DataFrame) -> List[str]:
    cols = set()
    for row in df.itertuples():
        cols |= set(row.column.split(";"))
    return sorted(cols - {""})


def _to_json(row: Dict[str, Any]) -> Dict[str, Any]:
    """"""
    nrow = copy.deepcopy(row)
    for key, val in row.items():
        if isinstance(val, int):
            continue
        elif isinstance(val, float):
            if math.isnan(val):
                del nrow[key]
            continue
        else:
            val = str(val).strip()
        if val.upper() == "TRUE":
            nrow[key] = True
            continue
        elif val.upper() == "FALSE":
            nrow[key] = False
            continue
        elif val.upper() in ["NA", "NK", "N/A", "NAN", "N/K", "NONE", "-", ""]:
            del nrow[key]
            continue
        else:
            pass
        try:
            nrow[key] = int(val)
        except ValueError:
            try:
                nrow[key] = float(val)
            except ValueError:
                nrow[key] = str(val)
    return nrow


def schema(
    schema_path: str, pattern: str = "*.csv", mostly: float = 0.95
) -> Callable[[pd.DataFrame], List[WorkUnitResult]]:
    schema_path = Path(schema_path)  # type: ignore
    with schema_path.open() as fp:
        schema = json.load(fp)
        validator = fastjsonschema.compile(schema)

    def rule_schema(df: pd.DataFrame):
        valids: List[Tuple[bool, str, str]] = []
        for row in map(_to_json, df.to_dict(orient="records")):
            try:
                validator(row)
                valids.append((True, "", ""))
            except fastjsonschema.exceptions.JsonSchemaValueException as e:
                valids.append((False, e.message, ";".join(e.path[1:])))
        valid_data = pd.DataFrame(valids, columns=["is_valid", "reason", "column"])
        rows_success = valid_data.is_valid.sum()
        rows_fail = len(valid_data) - rows_success
        ratio_success = rows_success / (rows_success + rows_fail)
        reason_counts = valid_data[~valid_data.is_valid].reason.value_counts()
        res = [
            dict(
                rows_success=rows_success,
                rows_fail=rows_fail,
                rows=rows_success + rows_fail,
                ratio_success=ratio_success,
                success=bool(ratio_success >= mostly),
                mostly=mostly,
                rows_fail_idx=[i for i, v in enumerate(valids) if v[0] is False],
                fail_data=None,
            )
        ]
        for reason, count in zip(reason_counts.index, reason_counts):
            res.append(
                dict(
                    rows=count,
                    rows_success=0,
                    rows_fail=count,
                    ratio_success=0,
                    success=False,
                    mostly=0,
                    rows_fail_idx=list(
                        valid_data.loc[valid_data.reason == reason].index
                    ),
                    reason=reason,
                    fail_data=json.dumps(
                        df.loc[valid_data.reason == reason][
                            _get_columns_from_valid_data(
                                valid_data[valid_data.reason == reason]
                            )
                        ]
                        .head(10)
                        .to_dict()
                    ),
                )
            )
        return res

    rule_schema.__doc__ = f"{schema.get('title', schema_path)} schema"
    rule_schema.__name__ = "schema_" + schema_path.stem.split(".")[0]
    rule_schema.pattern = pattern
    return rule_schema


def main(args=None):
    from .runner import _main

    _main(args)
