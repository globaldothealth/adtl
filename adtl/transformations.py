""" Functions which can be applied to source fields, allowing extensibility """

import logging
from datetime import datetime, timedelta, date

try:
    import zoneinfo
except ImportError:
    from backports import zoneinfo  # noqa

import pint


def isNotNull(value):
    return value not in [None, ""]


def textIfNotNull(field, return_val):
    return return_val if field not in [None, ""] else None


def getFloat(value):
    if not value:
        return None

    if '"' in value or " " in value:
        value = value.strip('"').replace(" ", "")

    # handle comma decimal separator
    value_int, _, fraction = value.partition(",")
    if "." in fraction:  # comma was being used as a thousands separator
        value = value_int + fraction
    else:
        # replace full stops as they may be used for thousands separator, first
        # then use full stop as decimal separator
        value = value.replace(".", "").replace(",", ".")
    try:
        return float(value)
    except ValueError:
        return value


def yearsElapsed(birthdate, currentdate):
    if birthdate in [None, ""] or currentdate in [None, ""]:
        return None

    bd = datetime.strptime(birthdate, "%Y-%m-%d")
    cd = datetime.strptime(currentdate, "%Y-%m-%d")

    days = cd - bd
    return pint.Quantity(days.days, "days").to("years").m


def durationDays(symptomstartdate, currentdate):
    """
    Returns the number of days between two dates.
    Preferable to Y-M-D elapsed, as month length is ambiguous -
    can be anywhere between 28-31 days.
    """
    if symptomstartdate in [None, ""] or currentdate in [None, ""]:
        return None

    bd = datetime.strptime(symptomstartdate, "%Y-%m-%d")
    cd = datetime.strptime(currentdate, "%Y-%m-%d")

    days = cd - bd
    return days.days


def startDate(enddate, duration):
    """
    Retuns the start date in ISO format, given the end date and the duration.
    """
    if enddate in [None, ""] or duration in [None, ""]:
        return None

    ed = datetime.strptime(enddate, "%Y-%m-%d")

    sd = ed - timedelta(days=duration)

    return sd.strftime("%Y-%m-%d")


def endDate(startdate, duration):
    """
    Retuns the end date in ISO format, given the start date and the duration.
    """
    if startdate in [None, ""] or duration in [None, ""]:
        return None

    sd = datetime.strptime(startdate, "%Y-%m-%d")

    duration = float(duration)

    ed = sd + timedelta(days=duration)

    return ed.strftime("%Y-%m-%d")


def makeDate(year, month, day):
    if year in [None, ""] or month in [None, ""] or day in [None, ""]:
        return None
    try:
        year, month, day = int(year), int(month), int(day)
    except ValueError:
        logging.error(
            f"Error in casting to integer: year={year}, month={month}, day={day}"
        )
        return None
    try:
        return date(year, month, day).isoformat()
    except ValueError:
        logging.error(
            f"Could not construct date from: year={year}, month={month}, day={day}"
        )
        return None


def makeDateTimeFromSeconds(date, time_seconds, date_format, tzname):
    if date == "":
        return None
    try:
        t = datetime.strptime(date, date_format).replace(
            tzinfo=zoneinfo.ZoneInfo(tzname)
        )
    except ValueError:
        logging.error(
            f"Could not convert date {date!r} from date format {date_format!r}"
        )
        return None
    if time_seconds == "":
        return t.date().isoformat()  # return date only
    time_seconds = int(time_seconds)
    hour = time_seconds // 3600
    minute = (time_seconds % 3600) // 60
    return t.replace(hour=hour, minute=minute).isoformat()
