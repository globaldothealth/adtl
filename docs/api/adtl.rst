=====================
Module reference
=====================

.. autoclass:: adtl.Parser
  :members:
