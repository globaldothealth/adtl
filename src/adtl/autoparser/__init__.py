import argparse
import sys

try:
    from .dict_writer import DictWriter, create_dict, generate_descriptions
    from .make_toml import ParserGenerator, create_parser
    from .mapping import Mapper, create_mapping
except ImportError:  # pragma: no cover
    raise ImportError(
        "autoparser is not available. Import as 'adtl[autoparser]' to use."
    )

__all__ = [
    "DictWriter",
    "create_dict",
    "generate_descriptions",
    "Mapper",
    "create_mapping",
    "ParserGenerator",
    "create_parser",
]

from .dict_writer import main as make_dd_main
from .make_toml import main as make_toml_main
from .mapping import main as csv_mapping_main


def main():
    parser = argparse.ArgumentParser(
        description="adtl-autoparser: A tool for creating data dictionaries and parsers"
    )

    subparsers = parser.add_subparsers(title="subcommands", dest="subcommand")

    # Subcommand: create-dict
    parser_create_dict = subparsers.add_parser(
        "create-dict",
        help="Create a data dictionary from a dataset",
    )
    parser_create_dict.set_defaults(func=make_dd_main)

    # Subcommand: create-mapping
    parser_create_mapping = subparsers.add_parser(
        "create-mapping",
        help="Create initial CSV mapping from data dictionary (LLM key required)",
    )
    parser_create_mapping.set_defaults(func=csv_mapping_main)

    # Subcommand: create-parser
    parser_create_parser = subparsers.add_parser(
        "create-parser",
        help="Generate TOML parser from CSV mapping file",
    )
    parser_create_parser.set_defaults(func=make_toml_main)

    args, unknown_args = parser.parse_known_args()

    if args.subcommand is None:
        parser.print_help()
        sys.exit(1)

    # Call the appropriate function with remaining arguments
    args.func(unknown_args)
