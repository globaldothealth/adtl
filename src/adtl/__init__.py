import argparse
import importlib.metadata
import json

from adtl.parser import Parser
from adtl.python_interface import parse

__all__ = ["Parser", "parse"]

__version__ = importlib.metadata.version("adtl")


def main(argv=None):
    cmd = argparse.ArgumentParser(
        prog="adtl",
        description="Transforms and validates data into CSV given a specification",
    )
    cmd.add_argument(
        "spec",
        help="specification file to use",
    )
    cmd.add_argument("file", help="file to read in")
    cmd.add_argument(
        "-o", "--output", help="output file, if blank, writes to standard output"
    )
    cmd.add_argument(
        "--encoding", help="encoding input file is in", default="utf-8-sig"
    )
    cmd.add_argument(
        "--parquet", help="output file is in parquet format", action="store_true"
    )
    cmd.add_argument(
        "-q",
        "--quiet",
        help="quiet mode - decrease verbosity, disable progress bar",
        action="store_true",
    )
    cmd.add_argument("--save-report", help="save report in JSON format")
    cmd.add_argument(
        "--include-def",
        action="append",
        help="include external definition (TOML or JSON)",
    )
    cmd.add_argument("--version", action="version", version="%(prog)s " + __version__)
    args = cmd.parse_args(argv)
    include_defs = args.include_def or []
    spec = Parser(args.spec, include_defs=include_defs, quiet=args.quiet)

    # check for incompatible options
    if spec.header.get("returnUnmatched") and args.parquet:
        raise ValueError("returnUnmatched and parquet options are incompatible")

    # run adtl
    adtl_output = spec.parse(args.file, encoding=args.encoding)
    adtl_output.save(args.output or spec.name, "parquet" if args.parquet else "csv")
    if args.save_report:
        adtl_output.report.update(
            dict(
                encoding=args.encoding,
                include_defs=include_defs,
                file=args.file,
                parser=args.spec,
            )
        )
        with open(args.save_report, "w") as fp:
            json.dump(adtl_output.report, fp, sort_keys=True, indent=2)
    else:
        adtl_output.show_report()


if __name__ == "__main__":
    main()
