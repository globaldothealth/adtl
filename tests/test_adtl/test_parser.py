from __future__ import annotations

import json
import warnings
from pathlib import Path

import pytest
from pytest_unordered import unordered

import adtl.parser as parser

TEST_PARSERS_PATH = Path(__file__).parent / "parsers"
TEST_SOURCES_PATH = Path(__file__).parent / "sources"
TEST_SCHEMAS_PATH = Path(__file__).parent / "schemas"

# test the header info is being picked up


def test_get_date_fields():
    with (Path(__file__).parent / "parsers" / "test.schema.json").open() as fp:
        schema = json.load(fp)
        assert parser.get_date_fields(schema) == unordered(
            ["enrolment_date", "admission_date"]
        )


def test_default_date_format(snapshot):
    transformed_csv_data = (
        parser.Parser(TEST_PARSERS_PATH / "epoch.json")
        .parse(TEST_SOURCES_PATH / "epoch.csv")
        .write_csv("table")
    )
    assert transformed_csv_data == snapshot


def test_provide_empty_values_definition(snapshot):
    ps = parser.Parser(TEST_PARSERS_PATH / "emptyFields.json")

    transformed_csv_data = ps.parse(TEST_SOURCES_PATH / "emptyFields.csv").data[
        "observation"
    ]

    assert transformed_csv_data == snapshot


def test_skip_field_pattern_present(snapshot):
    transformed_csv_data = (
        parser.Parser(TEST_PARSERS_PATH / "skip_field.json")
        .parse(TEST_SOURCES_PATH / "skip_field_present.csv")
        .write_csv("table")
    )
    assert transformed_csv_data == snapshot


def test_skip_field_pattern_absent(snapshot):
    transformed_csv_data = (
        parser.Parser(TEST_PARSERS_PATH / "skip_field.json")
        .parse(TEST_SOURCES_PATH / "skip_field_absent.csv")
        .write_csv("table")
    )
    assert transformed_csv_data == snapshot


def test_make_fields_optional():
    with (TEST_SCHEMAS_PATH / "epoch-oneOf.schema.json").open() as fp:
        schema = json.load(fp)
    assert schema["required"] == ["epoch", "id", "text"]
    assert parser.make_fields_optional(schema, ["text"])["required"] == ["epoch", "id"]
    assert parser.make_fields_optional(schema, ["field_not_present"])["required"] == [
        "epoch",
        "id",
        "text",
    ]
    assert parser.make_fields_optional(schema, ["sex"])["oneOf"] == [
        {"required": []},
        {"required": ["sex_at_birth"]},
    ]
    assert "oneOf" not in parser.make_fields_optional(schema, ["sex", "sex_at_birth"])

    assert schema["anyOf"] == [
        {"required": ["sex", "epoch"]},
        {"required": ["sex_at_birth", "epoch"]},
    ]

    assert parser.make_fields_optional(schema, ["epoch"])["anyOf"] == [
        {"required": ["sex"]},
        {"required": ["sex_at_birth"]},
    ]
    assert parser.make_fields_optional(schema, ["sex", "sex_at_birth"])["anyOf"] == [
        {"required": ["epoch"]}
    ]


# Check writing out and validation works as expected


def test_parse_write_buffer(snapshot):
    groupby_source = [
        {
            "sex": "1",
            "subjid": "S007",
            "dsstdat": "2020-05-06",
            "hostdat": "2020-06-08",
        },
        {
            "sex": "2",
            "subjid": "S001",
            "dsstdat": "2022-01-11",
            "hostdat": "2020-06-08",
        },
    ]
    ps = parser.Parser(TEST_PARSERS_PATH / "groupBy.json")
    buf = ps.parse_rows(groupby_source, "test_groupby").write_csv("subject")
    assert buf == snapshot


def test_validation(snapshot):
    invalid_groupby_source = [
        {
            "sex": "1",
            "subjid": "S007",
            "dsstdat": "2020-05-06",
            "hostdat": "2020-06-08",
            "ethnic": "1",
        },
        {
            "sex": "",
            "subjid": "S007",
            "dsstdat": "",
            "hostdat": "",
            "ethnic": "",
        },
        {
            "sex": "5",
            "subjid": "S001",
            "dsstdat": "2022-01-11",
            "hostdat": "8/6/2022",
            "ethnic": "2",
        },
        {
            "sex": "1",
            "subjid": "S009",
            "dsstdat": "2020-05-06",
            "hostdat": "8/6/2020",
            "ethnic": "3",
        },
    ]

    ps = parser.Parser(TEST_PARSERS_PATH / "groupBy-with-schema.json")
    buf = ps.parse_rows(invalid_groupby_source, "test_groupby_invalid").write_csv(
        "subject"
    )
    assert buf == snapshot


# check static tables


def test_constant_table():
    ps = parser.Parser(TEST_PARSERS_PATH / "constant.json").parse_rows([{"x": 1}], "x")
    assert list(ps.read_table("metadata")) == [
        {"dataset": "constant", "version": "20220505.1", "format": "csv"}
    ]


# kind=groupby behaviour


def test_multi_id_groupby(snapshot):
    multi_id_groupby = [
        {
            "sex": "1",
            "subjid": "",
            "othid": "P007",
            "dsstdat": "2020-05-06",
            "hostdat": "2020-06-08",
        },
        {
            "sex": "2",
            "subjid": "S001",
            "othid": "P008",
            "dsstdat": "2022-01-11",
            "hostdat": "2020-06-08",
        },
    ]

    ps = parser.Parser(TEST_PARSERS_PATH / "groupBy-multi-id.json")
    buf = ps.parse_rows(multi_id_groupby, "groupby_multi_id").write_csv("subject")
    assert buf == snapshot


def test_apply_when_values_are_present_groupby():
    data = [
        {
            "subjid": "S007",
            "brthdtc": "1996-02-24",
            "dsstdat": "2023-02-24",
            "age": "22",
            "ageu": 1,
            "icu_hostdat": 1,
        }
    ]

    apply_values_present_output = list(
        parser.Parser(TEST_PARSERS_PATH / "apply.toml")
        .parse_rows(data, "apply")
        .read_table("subject")
    )

    assert apply_values_present_output == [
        {
            "subject_id": "S007",
            "age": pytest.approx(27.0, 0.001),
            "icu_admitted": True,
            "dob_year": 1974,
        }
    ]


def test_apply_when_values_not_present_groupby():
    data = [
        {
            "subjid": "S007",
            "brthdtc": "",
            "dsstdat": "2023-02-24",
            "age": "22",
            "ageu": 1,
            "icu_hostdat": "",
        }
    ]
    apply_values_absent_output = list(
        parser.Parser(TEST_PARSERS_PATH / "apply.toml")
        .parse_rows(data, "apply_absent")
        .read_table("subject")
    )

    assert apply_values_absent_output == [
        {"subject_id": "S007", "age": 22.0, "icu_admitted": False, "dob_year": 2001}
    ]


def test_no_overwriting_groupby():
    prsr = parser.Parser(TEST_PARSERS_PATH / "stop-overwriting.toml")

    overwriting_output = list(
        prsr.parse(TEST_SOURCES_PATH / "stop-overwriting.csv").read_table("visit")
    )
    # Can't use snapshot as the antiviral order changes between runs
    assert overwriting_output == [
        {
            "subject_id": 1,
            "earliest_admission": "2023-11-19",
            "start_date": "2023-11-20",
            "treatment_antiviral_type": unordered(["Ribavirin", "Interferon"]),
        },
        {
            "subject_id": 2,
            "start_date": "2022-11-23",
            "icu_admission_date": unordered(["2020-11-25", "2020-11-30"]),
            "treatment_antiviral_type": ["Lopinavir"],
        },
        {
            "subject_id": 3,
            "start_date": "2020-02-20",
            "treatment_antiviral_type": unordered(
                ["Ribavirin", "Lopinavir", "Interferon"]
            ),
        },
    ]


@pytest.mark.parametrize(
    "verbosity,expected_warnings",
    [
        (False, None),
        (True, "Multiple rows of data found for"),
    ],
)
def test_overwriting_with_strict_groupby(verbosity, expected_warnings):
    prsr = parser.Parser(TEST_PARSERS_PATH / "stop-overwriting.toml", verbose=verbosity)
    prsr.tables["visit"]["aggregation"] = "lastNotNullStrict"

    if verbosity:
        with pytest.warns(UserWarning, match=expected_warnings):
            overwritten_output = list(
                prsr.parse(TEST_SOURCES_PATH / "stop-overwriting.csv").read_table(
                    "visit"
                )
            )
    else:
        warnings.filterwarnings("error")  # Treat warnings as errors
        overwritten_output = list(
            prsr.parse(TEST_SOURCES_PATH / "stop-overwriting.csv").read_table("visit")
        )

    assert overwritten_output == [
        {
            "subject_id": 1,
            "earliest_admission": "2023-11-19",
            "start_date": "2023-11-19",
            "treatment_antiviral_type": unordered(["Ribavirin"]),
        },
        {
            "subject_id": 2,
            "start_date": "2020-11-23",
            "icu_admission_date": unordered(["2020-11-30"]),
            "treatment_antiviral_type": ["Lopinavir"],
        },
        {
            "subject_id": 3,
            "start_date": "2020-02-20",
            "treatment_antiviral_type": unordered(["Ribavirin"]),
        },
    ]


# General tests


def test_parser_clear_removes_all_data():
    ps = parser.Parser(TEST_PARSERS_PATH / "oneToMany.json")
    ps.data = {"observation": []}
    ps.clear()
    assert ps.data == {}


def test_read_table_raises_error_if_wrong_table_name():
    source = [
        {
            "Entry_ID": 1,
            "Epoch": "11/01/1999",
            "SomeDate": "24/01/1999",
            "Text": "Lorem Ipsum",
        }
    ]
    with pytest.raises(ValueError, match="Invalid table name"):
        list(
            parser.Parser(TEST_PARSERS_PATH / "epoch.json")
            .parse_rows(source, "one_to_many")
            .read_table("wrong-name")
        )


@pytest.mark.filterwarnings("ignore:No matches found")
@pytest.mark.filterwarnings("ignore:Could not construct date")
def test_return_unmapped(snapshot):
    transformed_csv_data = (
        parser.Parser(TEST_PARSERS_PATH / "return-unmapped.toml")
        .parse(TEST_SOURCES_PATH / "return-unmapped.csv")
        .write_csv("subject")
    )
    assert transformed_csv_data == snapshot


def test_subschema_validation_for_large_schemas(snapshot):
    transformed_csv_data = (
        parser.Parser(TEST_PARSERS_PATH / "long-oneof-parser.toml")
        .parse(TEST_SOURCES_PATH / "long-oneof.csv")
        .data["long"]
    )
    assert transformed_csv_data == snapshot
