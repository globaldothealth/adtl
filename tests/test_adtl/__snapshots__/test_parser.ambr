# serializer version: 1
# name: test_default_date_format
  '''
  adtl_valid,adtl_error,epoch,id,some_date,text
  True,,1999-01-11,1,1999-01-24,Lorem ipsum
  True,,2022-12-19,2,2023-01-27,example
  
  '''
# ---
# name: test_multi_id_groupby
  '''
  admission_date,country_iso3,dataset_id,enrolment_date,sex_at_birth,subject_id
  2020-06-08,GBR,dataset-2020-03-23,2020-05-06,male,P007
  2020-06-08,GBR,dataset-2020-03-23,2022-01-11,female,S001
  
  '''
# ---
# name: test_parse_write_buffer
  '''
  admission_date,country_iso3,dataset_id,enrolment_date,sex_at_birth,subject_id
  2020-06-08,GBR,dataset-2020-03-23,2020-05-06,male,S007
  2020-06-08,GBR,dataset-2020-03-23,2022-01-11,female,S001
  
  '''
# ---
# name: test_provide_empty_values_definition
  list([
    dict({
      'is_present': True,
      'name': 'headache',
    }),
  ])
# ---
# name: test_return_unmapped
  '''
  age,date_death,date_of_birth,diabetes_type,has_smoking,pregnancy_birth_weight_kg,subject_id
  fifteen,2023,2023-11-20,no diabetes,today,eight,1
  
  '''
# ---
# name: test_skip_field_pattern_absent
  '''
  adtl_valid,adtl_error,cough,epoch,followup_cough,headache,id,text
  False,data.epoch must be date,1,11/01/1999,,,1,Lorem ipsum
  False,data.epoch must be date,0,19/12/2022,,,2,example
  
  '''
# ---
# name: test_skip_field_pattern_present
  '''
  adtl_valid,adtl_error,cough,epoch,followup_cough,headache,id,text
  False,data.epoch must be date,1,11/01/1999,0,3,1,Lorem ipsum
  False,data.epoch must be date,0,19/12/2022,1,0,2,example
  
  '''
# ---
# name: test_subschema_validation_for_large_schemas
  list([
    dict({
      'adtl_valid': True,
      'attribute': 'lesion_ml',
      'phase': 'presentation',
      'subjid': '1',
      'value': '2-5',
    }),
    dict({
      'adtl_valid': True,
      'attribute': 'comor_dementia',
      'phase': 'presentation',
      'subjid': '1',
      'value': 'Yes',
    }),
    dict({
      'adtl_valid': True,
      'attribute': 'vital_rass',
      'phase': 'presentation',
      'subjid': '1',
      'value_num': 1.0,
    }),
    dict({
      'adtl_error': "data.value must be one of ['None', '1', '2-5', '6-9', '10-24', '25-49', '50-99', '100-250', '251-1000', '>1000']",
      'adtl_valid': False,
      'attribute': 'lesion_ml',
      'phase': 'presentation',
      'subjid': '2',
      'value': 'five',
    }),
    dict({
      'adtl_valid': True,
      'attribute': 'comor_dementia',
      'phase': 'presentation',
      'subjid': '2',
      'value': 'Yes',
    }),
    dict({
      'adtl_valid': True,
      'attribute': 'vital_rass',
      'phase': 'presentation',
      'subjid': '2',
      'value_num': 2.0,
    }),
    dict({
      'adtl_valid': True,
      'attribute': 'lesion_ml',
      'phase': 'presentation',
      'subjid': '3',
      'value': '2-5',
    }),
    dict({
      'adtl_valid': True,
      'attribute': 'comor_dementia',
      'phase': 'presentation',
      'subjid': '3',
      'value': 'Yes',
    }),
    dict({
      'adtl_error': "data must contain ['subjid', 'phase', 'attribute', 'value'] properties",
      'adtl_valid': False,
      'attribute': 'comor_asthma',
      'phase': 'presentation',
      'subjid': '3',
      'value_num': 'False',
    }),
    dict({
      'adtl_valid': True,
      'attribute': 'vital_rass',
      'phase': 'presentation',
      'subjid': '3',
      'value_num': 3.0,
    }),
  ])
# ---
# name: test_validation
  '''
  adtl_valid,adtl_error,admission_date,country_iso3,dataset_id,enrolment_date,ethnicity,sex_at_birth,subject_id
  True,,2020-06-08,GBR,dataset-2020-03-23,2020-05-06,['Arabic'],male,S007
  False,"data must contain ['subject_id', 'country_iso3', 'enrolment_date', 'sex_at_birth'] properties",8/6/2022,GBR,dataset-2020-03-23,2022-01-11,['Black'],,S001
  False,data.admission_date must be date,8/6/2020,GBR,dataset-2020-03-23,2020-05-06,['South_Asian'],male,S009
  
  '''
# ---
